plugins/BEdita/Core/src/Model/Table/FoldersTable.php

```diff
@@ -0,0 +1,153 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2018 ChannelWeb Srl, Chialab Srl
+ *
+ * This file is part of BEdita: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+
+namespace BEdita\Core\Model\Table;
+
+use BEdita\Core\Model\Entity\Folder;
+use Cake\Datasource\EntityInterface;
+use Cake\Event\Event;
+use Cake\ORM\RulesChecker;
+use Cake\ORM\Rule\ValidCount;
+use Cake\ORM\TableRegistry;
+
+/**
+ * Folders Model
+ *
+ * @property \Cake\ORM\Association\HasOne $Trees
+ * @property \Cake\ORM\Association\BelongsToMany $Children
+ *
+ * @method \BEdita\Core\Model\Entity\Folder get($primaryKey, $options = [])
+ * @method \BEdita\Core\Model\Entity\Folder newEntity($data = null, array $options = [])
+ * @method \BEdita\Core\Model\Entity\Folder[] newEntities(array $data, array $options = [])
+ * @method \BEdita\Core\Model\Entity\Folder|bool save(\Cake\Datasource\EntityInterface $entity, $options = [])
+ * @method \BEdita\Core\Model\Entity\Folder patchEntity(\Cake\Datasource\EntityInterface $entity, array $data, array $options = [])
+ * @method \BEdita\Core\Model\Entity\Folder[] patchEntities($entities, array $data, array $options = [])
+ * @method \BEdita\Core\Model\Entity\Folder findOrCreate($search, callable $callback = null, $options = [])
+ *
+ * @since 4.0.0
+ *
+ */
+class FoldersTable extends ObjectsTable
+{
+
+    /**
+     * {@inheritDoc}
+     *
+     * @codeCoverageIgnore
+     */
+    public function initialize(array $config)
+    {
+        parent::initialize($config);
+
+        $this->setEntityClass(Folder::class);
+
+        $this->belongsToMany('Children', [
+            'className' => 'Objects',
+            'through' => 'Trees',
+            'foreignKey' => 'parent_id',
+            'targetForeignKey' => 'object_id',
+            'sort' => [
+                'Trees.tree_left' => 'asc',
+            ],
+        ]);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @codeCoverageIgnore
+     */
+    public function buildRules(RulesChecker $rules)
+    {
+        $rules->add(
+            [$this, 'hasAtMostOneParent'],
+            'hasAtMostOneParent',
+            [
+                'errorField' => 'parents',
+                'message' => __d('bedita', 'Folder can have at most one existing parent.'),
+            ]
+        );
+
+        return $rules;
+    }
+
+    /**
+     * Custom rule for checking that entity has at most one parent.
+     * The check is done on `parents` property
+     *
+     * @param EntityInterface $entity The entity to check
+     * @return bool
+     */
+    public function hasAtMostOneParent(EntityInterface $entity)
+    {
+        if (empty($entity->parents)) {
```

Comment:
Typical `Scrutinizer` complaint:
should we use `$entity->get('parents')` here instead?


```
Accessing parents on the interface Cake\Datasource\EntityInterface suggest that you code against a concrete implementation. How about adding an instanceof check?```